clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
-    // Create a parameter descriptor and append it to the result
-    ParamDescs.push_back(makeParamDesc(Fld, ArgType));
-  };
-
-  // Creates a parameter descriptor for SYCL special object - SYCL accessor or
-  // sampler.
+// The first template argument to the kernel function is used to identify the
+// kernel itself.
+static QualType calculateKernelNameType(ASTContext &Ctx,
+                                        FunctionDecl *KernelCallerFunc) {
+  // TODO: Not sure what the 'fully qualified type's purpose is here, the type
+  // itself should have its full qualified name, so figure out what the purpose
+  // is.
+  const TemplateArgumentList *TAL =
+      KernelCallerFunc->getTemplateSpecializationArgs();
+  return TypeName::getFullyQualifiedType(TAL->get(0).getAsType(), Ctx,
+                                         /*WithGlobalNSPrefix=*/true);
+}
+
+// Gets a name for the kernel caller func, calculated from the first template
+// argument.
+static std::string constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
+                                       MangleContext &MC, bool StableName) {
+  QualType KernelNameType =
+      calculateKernelNameType(S.getASTContext(), KernelCallerFunc);
+
+  if (StableName)
+    return PredefinedExpr::ComputeName(S.getASTContext(),
+                                       PredefinedExpr::UniqueStableNameType,
+                                       KernelNameType);
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+
+  MC.mangleTypeName(KernelNameType, Out);
+  return std::string(Out.str());
+}
+
+// anonymous namespace so these don't get linkage.
+namespace {
+
+QualType getItemType(const FieldDecl *FD) { return FD->getType(); }
+QualType getItemType(const CXXBaseSpecifier &BS) { return BS.getType(); }
+
+// Implements the 'for-each-visitor'  pattern.
+template <typename ParentTy, typename... Handlers>
+static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
+                                 CXXRecordDecl *Wrapper,
+                                 Handlers &... handlers);
+
+template <typename RangeTy, typename... Handlers>
+static void VisitAccessorWrapperHelper(CXXRecordDecl *Owner, RangeTy Range,
+                                       Handlers &... handlers) {
+  for (const auto &Item : Range) {
+    QualType ItemTy = getItemType(Item);
+    if (Util::isSyclAccessorType(ItemTy))
+      (void)std::initializer_list<int>{
+          (handlers.handleSyclAccessorType(Item, ItemTy), 0)...};
+    else if (ItemTy->isStructureOrClassType()) {
+      VisitAccessorWrapper(Owner, Item, ItemTy->getAsCXXRecordDecl(),
+                           handlers...);
+      if (Util::isSyclStreamType(ItemTy))
+        (void)std::initializer_list<int>{
+            (handlers.handleSyclStreamType(Item, ItemTy), 0)...};
+    }
+  }
+}
+
+// poorly named Parent is the 'how we got here', basically just enough info for
```

Comment:
Replace comment with:

Parent is contains the FieldDecl or CXXBaseSpecifier that was used to enter the Wrapper structure that we're currently visiting.  Owner is the parent type (which doesn't exist in cases where it is a FieldDecl in the 'root'), and Wrapper is the current struct being unwrapped.

Feel free to wordsmith/clarify as you can.